llvm/lib/Target/BPF/BPFISelLowering.cpp

```diff
@@ -481,8 +495,47 @@ SDValue BPFTargetLowering::LowerFormalArguments(
 
   if (Subtarget->isSolana()) {
     if (IsVarArg) {
-      fail(DL, DAG, "Functions with VarArgs are not supported");
-      assert(false);
+      ArrayRef<MCPhysReg> ArgRegs = makeArrayRef(SBFArgRegs);
+      unsigned Idx = CCInfo.getFirstUnallocated(ArgRegs);
+      const TargetRegisterClass *RC = getRegClassFor(MVT::i64);
+
+      // Offset of the first variable argument from stack pointer.
+      int VaArgOffset;
+
+      if (ArgRegs.size() == Idx)
+        VaArgOffset = alignTo(CCInfo.getNextStackOffset(), 8);
+      else {
+        VaArgOffset = -((int)(8 * (ArgRegs.size() - Idx)));
+      }
+
+      // Record the frame index of the first variable argument
+      // which is a value necessary to VASTART.
+      int FI = MFI.CreateFixedObject(8, VaArgOffset, true);
+      BFI->setVarArgsFrameIndex(FI);
+
+      // Used with vargs to acumulate store chains.
+      std::vector<SDValue> OutChains;
+
+      for (unsigned I = Idx; I < ArgRegs.size();
+          ++I, VaArgOffset += 8) {
```

Comment:
Should probably explain what this loop does as the code is completely unreadable.